### PR TITLE
[objc] Add NSDecimalNumber <-> decimal support

### PR DESCRIPTION
**Pull request by dalexsoto**
Work in progress

- [x] ReturnType Support
- [x] Array ReturnType Support
- [x] Make iOS work (linker.xml)
- [x] Parameter Support
- [x] Array Parameter Support

### CODE REVIEW

**Comment by dalexsoto** [comment]
Feedback is welcome 😄

File: objcgen/objcgenerator.cs [hunk 1]
```diff
@@ -1010,6 +1012,24 @@ void ReturnValue (Type t)
...
+				implementation.WriteLine ("MonoString* __decimalmonostr = (MonoString *) mono_runtime_invoke (__tostringmethod, __unboxedresult, __tostringargs, &__ex);");
+				implementation.WriteLine ("if (__ex)");
+				implementation.Indent++;
+				implementation.WriteLine ("mono_embeddinator_throw_exception (__ex);");
+				implementation.Indent--;
+				implementation.WriteLine ("NSString* __decimalnsstr = mono_embeddinator_get_nsstring (__decimalmonostr);");
+				// Force NSDecimalNumber to parse the number using dot as decimal separator http://stackoverflow.com/a/7905775/572076
+				implementation.WriteLine ("NSDictionary* __forcedotseparator = @{ NSLocaleDecimalSeparator : @\".\" };");
+				implementation.WriteLine ("NSDecimalNumber* __nsdecresult = [NSDecimalNumber decimalNumberWithString:__decimalnsstr locale:__forcedotseparator];");
+				implementation.WriteLine ("return __nsdecresult;");
```
Comment (rolfbjarne) [inline objcgen/objcgenerator.cs]: You can put this whole chunk of code into a method: `mono_embeddinator_get_nsdecimalnumber (MonoObject *obj)`, which would simplify and minimize the generated code a lot (a single method call).

File: tests/managed/types.cs [hunk 2]
```diff
@@ -87,3 +87,9 @@ public class MyException : Exception {
 // objc: this type won't be generated (subclassing an unsupported type) but the generation will succeed (with warnings)
 public class MyNextException : MyException {
 }
+
+public static class Type_Decimal {
+	public static decimal Max { get { return decimal.MaxValue; } }
+	public static decimal Min { get { return decimal.MinValue; } }
+	public static decimal Zero { get { return decimal.Zero; } }
```
Comment (rolfbjarne) [inline tests/managed/types.cs]: Add a few tests for numbers with decimal points as well.

File: objcgen/objcgenerator.cs [hunk 3]
```diff
@@ -1010,6 +1012,24 @@ void ReturnValue (Type t)
 			case TypeCode.String:
 				implementation.WriteLine ("return mono_embeddinator_get_nsstring ((MonoString *) __result);");
 				break;
+			case TypeCode.Decimal:
+				implementation.WriteLine ("void* __unboxedresult = mono_object_unbox (__result);");
+				implementation.WriteLine ("MonoObject* __invariantculture = mono_embeddinator_get_cultureinfo_invariantculture_object ();");
+				implementation.WriteLine ($"MonoMethod* __tostringmethod = mono_embeddinator_lookup_method (\":ToString(System.IFormatProvider)\", mono_object_get_class (__result));");
```
Comment (spouliot) [inline objcgen/objcgenerator.cs]: there should be a `TOKENLOOKUP` alternative

Comment (spouliot) [inline objcgen/objcgenerator.cs]: won't be possible by moving code, but it's cached (and better not duplicated)

File: objcgen/objcgenerator.cs [hunk 4]
```diff
@@ -1010,6 +1012,24 @@ void ReturnValue (Type t)
...
+				implementation.WriteLine ("__tostringargs [0] = __invariantculture;");
+				implementation.WriteLine ("MonoObject* __ex = nil;");
+				implementation.WriteLine ("MonoString* __decimalmonostr = (MonoString *) mono_runtime_invoke (__tostringmethod, __unboxedresult, __tostringargs, &__ex);");
+				implementation.WriteLine ("if (__ex)");
+				implementation.Indent++;
+				implementation.WriteLine ("mono_embeddinator_throw_exception (__ex);");
+				implementation.Indent--;
+				implementation.WriteLine ("NSString* __decimalnsstr = mono_embeddinator_get_nsstring (__decimalmonostr);");
+				// Force NSDecimalNumber to parse the number using dot as decimal separator http://stackoverflow.com/a/7905775/572076
+				implementation.WriteLine ("NSDictionary* __forcedotseparator = @{ NSLocaleDecimalSeparator : @\".\" };");
```
Comment (spouliot) [inline objcgen/objcgenerator.cs]: might be cheaper to keep cached than re-created at each call. No cost if you don't use `decimal` but if you use it it's likely more than once.

File: tests/objc-cli/libmanaged/Tests/Tests.m [hunk 5]
```diff
@@ -413,6 +413,18 @@ - (void)testTypes {
     XCTAssertEqualObjects ((__bridge NSString *) nil, [Type_String nullString], "null string");
     XCTAssertEqualObjects (@"", [Type_String emptyString], "empty string");
     XCTAssertEqualObjects (@"Hello World", [Type_String nonEmptyString], "non-empty string");
+
+	NSDecimalNumber * decimalzero = [Type_Decimal zero];
+	NSDecimalNumber * nsdecimalzero = [NSDecimalNumber zero];
+	XCTAssertEqualObjects (nsdecimalzero, decimalzero, "NSDecimalNumber zero");
+
+	NSDecimalNumber * decimalmin = [Type_Decimal min];
```
Comment (chamons) [inline tests/objc-cli/libmanaged/Tests/Tests.m]: Maybe a few non-boundary cases as well? -1, 1, 42 along with some with with fractional parts.

Also, you have tabs while above has spaces. Maybe fix rest of test method?

Comment (spouliot) [inline tests/objc-cli/libmanaged/Tests/Tests.m]: please only fix spaces (to tabs) in a separate PR to ease reviewing :)

File: support/linker.xml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+﻿<?xml version="1.0" encoding="UTF-8" ?>
+<linker>
+    <assembly fullname="mscorlib">
+        <type fullname="System.Decimal">
+            <method signature="System.String ToString(System.IFormatProvider)" />
+		</type>
```
Comment (spouliot) [inline support/linker.xml]: ident

File: objcgen/driver.cs [hunk 7]
```diff
@@ -521,6 +521,7 @@ public int Compile ()
 						mtouch.Append ("--msym:false ");
 						mtouch.Append ("--nosymbolstrip ");
 						mtouch.Append ($"--embeddinator ");
+						mtouch.Append ($"--xml={Path.Combine (OutputDirectory, "linker.xml")} ");
```
Comment (spouliot) [inline objcgen/driver.cs]: The idea was to generate the XML file depending on the assemblies code, e.g. `System.Decimal` is not always needed (it's even uncommon) so we don't want all project to pay the price for it.

However this can be used until the correct support is added, just add a `FIXME` until then.

Comment (dalexsoto) [inline objcgen/driver.cs]: Yep I should have noted this as a workaround since I guess we also need to accept a `linker.xml` from the user and merge our definitions. This was more to get my testing on iOS going 😃

File: support/mono_embeddinator.h [hunk 8]
```diff
@@ -185,4 +185,51 @@ void mono_embeddinator_init_object(MonoEmbedObject* object, MonoObject* instance
...
+MONO_EMBEDDINATOR_API
+MonoObject* mono_embeddinator_get_cultureinfo_invariantculture_object ();
+
+/**
+ * Gets decimal MonoClass.
+ */
+MONO_EMBEDDINATOR_API
+MonoClass* mono_embeddinator_get_decimal_class ();
+
+// from: https://github.com/mono/mono/blob/master/mono/metadata/decimal-ms.h
```
Comment (rolfbjarne) [inline support/mono_embeddinator.h]: This struct should go in `mono-support.h`.

File: support/objc-support.m [hunk 9]
```diff
@@ -61,3 +61,29 @@ NSComparisonResult mono_embeddinator_compare_to (MonoEmbedObject *object, MonoMe
 	int gchandle = (int) [native performSelector:@selector (xamarinGetGCHandle)];
 	return mono_gchandle_get_target (gchandle);
 }
+
+NSDecimalNumber* mono_embeddinator_get_nsdecimalnumber (void* __unboxedresult)
+{
+	static NSDictionary* __forcedotseparator = nil;
```
Comment (rolfbjarne) [inline support/objc-support.m]: You don't need all the ugly underscores in the variable names when the code isn't generated :smile:

**Comment by dalexsoto** [comment]
@spouliot / @rolfbjarne Ready for final review 👍

File: objcgen/objcgenerator.cs [hunk 10]
```diff
@@ -562,6 +562,18 @@ void GenerateArgumentArray (string parameterName, string argumentName, Type t, r
 				implementation.WriteLine ("else");
 				implementation.Indent++;
 				implementation.WriteLine ($"mono_array_set ({pnameArr}, {typeName}, {pnameIdx}, {pnameRet}.{returnValue});");
+				implementation.Indent--;
```
Comment (spouliot) [inline objcgen/objcgenerator.cs]: That's out of code change. Is that fixing the current indentation issues ? (noticed one earlier today)

File: objcgen/objcgenerator.cs [hunk 11]
```diff
@@ -603,6 +615,7 @@ void GenerateArgumentArray (string parameterName, string argumentName, Type t, r
 					implementation.WriteLine ($"mono_array_set ({pnameArr}, MonoObject *, {pnameIdx}, mono_gchandle_get_target ({pnameRet}->_object->_handle));");
 				else if (hasProtocol)
 					implementation.WriteLine ($"mono_array_set ({pnameArr}, MonoObject *, {pnameIdx}, mono_embeddinator_get_object ({pnameRet}, true));");
+				implementation.Indent--;
```
Comment (spouliot) [inline objcgen/objcgenerator.cs]: same as above

File: objcgen/objcgenerator.cs [hunk 12]
```diff
@@ -562,6 +562,18 @@ void GenerateArgumentArray (string parameterName, string argumentName, Type t, r
...
 				implementation.WriteLine ($"mono_array_set ({pnameArr}, {typeName}, {pnameIdx}, {pnameRet}.{returnValue});");
+				implementation.Indent--;
+				break;
+			case TypeCode.Decimal:
+				implementation.WriteLine ($"NSDecimalNumber* {pnameRet} = {parameterName}[{pnameIdx}];");
+				implementation.WriteLine ($"if (!{pnameRet} || [{pnameRet} isKindOfClass:[NSNull class]])");
+				implementation.Indent++;
+				implementation.WriteLine ($"continue;");
+				implementation.Indent--;
+				implementation.WriteLine ("else");
```
Comment (spouliot) [inline objcgen/objcgenerator.cs]: do you need the `else` after the `continue` ?
if not then we can reduce the code (and indentation level)

File: support/objc-support.h [hunk 13]
```diff
@@ -45,5 +45,11 @@ MonoObject* mono_embeddinator_get_object (id native, bool assertOnFailure);
 MONO_EMBEDDINATOR_API
 MonoAssembly *mono_embeddinator_find_assembly_in_bundle (const char *assembly);
 
+MONO_EMBEDDINATOR_API
+NSDecimalNumber* mono_embeddinator_get_nsdecimalnumber (void* __unboxedresult);
+
+MONO_EMBEDDINATOR_API
+MonoDecimal mono_embeddinator_get_monodecimal (NSDecimalNumber* nsdecimalnumber, mono_embeddinator_context_t* context);
```
Comment (spouliot) [inline support/objc-support.h]: `mono` one time too many ;-) maybe just `mono_embeddinator_get_decimal` ? or `mono_embeddinator_get_system_decimal`

File: objcgen/driver.cs [hunk 14]
```diff
@@ -549,6 +549,9 @@ public int Compile ()
 						mtouch.Append ("--dsym:false ");
 						mtouch.Append ("--msym:false ");
 						mtouch.Append ($"--embeddinator ");
+						// FIXME: we need to generate this XML file, issue #301
+						// https://github.com/mono/Embeddinator-4000/issues/301
+						mtouch.Append ($"--xml={Path.Combine (OutputDirectory, "linker.xml")} ");
```
Comment (spouliot) [inline objcgen/driver.cs]: ^ needs a call to `Quote` or it will fails if `OutputDirectory` has any space in it

File: objcgen/objcgenerator.cs (hunk 10 above)
Comment (dalexsoto) [inline objcgen/objcgenerator.cs]: Yes, this fixes indentation

File: objcgen/objcgenerator.cs (hunk 11 above)
Comment (dalexsoto) [inline objcgen/objcgenerator.cs]: Yes, this fixes indentation issues

**Comment by dalexsoto** [comment]
@spouliot fixed 😄

**Comment by spouliot** [comment]
@dalexsoto now you'll need to resolve your conflicts ;-)

**Comment by dalexsoto** [comment]
@spouliot already resolved 😎
